cpp/log/cert.cc

```diff
@@ -1419,49 +1429,53 @@ void CertChain::ClearChain() {
 }
 
 
-Cert::Status PreCertChain::UsesPrecertSigningCertificate() const {
+util::StatusOr<bool> PreCertChain::UsesPrecertSigningCertificate() const {
   const Cert* issuer = PrecertIssuingCert();
   if (!issuer) {
     // No issuer, so it must be a real root CA from the store.
-    return Cert::FALSE;
+    return false;
   }
 
-  return StatusOrBoolToCertStatus(
-      issuer->HasExtendedKeyUsage(cert_trans::NID_ctPrecertificateSigning));
+  return issuer->HasExtendedKeyUsage(cert_trans::NID_ctPrecertificateSigning);
 }
 
 
-Cert::Status PreCertChain::IsWellFormed() const {
+util::StatusOr<bool> PreCertChain::IsWellFormed() const {
   if (!IsLoaded()) {
     LOG(ERROR) << "Chain is not loaded";
-    return Cert::ERROR;
+    return util::Status(Code::FAILED_PRECONDITION, "Cert not loaded");
   }
 
   const Cert* pre = PreCert();
 
   // (1) Check that the leaf contains the critical poison extension.
-  Cert::Status status = pre->HasCriticalExtension(cert_trans::NID_ctPoison);
-  if (status != Cert::TRUE)
+  StatusOr<bool> status = pre->HasCriticalExtension(cert_trans::NID_ctPoison);
+  if (!status.ok() || !status.ValueOrDie()) {
     return status;
+  }
 
   // (2) If signed by a Precertificate Signing Certificate, check that
   // the AKID extensions are compatible.
-  status = UsesPrecertSigningCertificate();
-  if (status == Cert::FALSE) {
+  StatusOr<bool> uses_precert_signing = UsesPrecertSigningCertificate();
+  if (uses_precert_signing.ok() && !uses_precert_signing.ValueOrDie()) {
     // If there is no precert signing extendedKeyUsage, no more checks:
     // the cert was issued by a regular CA.
-    return Cert::TRUE;
+    return true;
   }
-  if (status != Cert::TRUE)
+  if (!uses_precert_signing.ok() || !uses_precert_signing.ValueOrDie()) {
     return status;
+  }
 
   const Cert* issuer = PrecertIssuingCert();
   // If pre has the extension set but the issuer doesn't, error.
-  status = pre->HasExtension(NID_authority_key_identifier);
-  if (status == Cert::FALSE)
-    return Cert::TRUE;
-  if (status != Cert::TRUE)
-    return status;
+  StatusOr<bool> has_akid = pre->HasExtension(NID_authority_key_identifier);
+
+  if (has_akid.ok() && !has_akid.ValueOrDie()) {
+    return true;
+  }
+  if (!has_akid.ok()) {
+    return has_akid;
+  }
```

Comment:
Could add `CHECK(has_akid.ValueOrDie())` here to guard against future modifications to this getting it wrong/document the invariant.